src/cpp/core/file_lock/LinkBasedFileLock.cpp

```diff
@@ -0,0 +1,318 @@
+/*
+ * LinkBasedFileLock.cpp
+ *
+ * Copyright (C) 2009-12 by RStudio, Inc.
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+#include <core/FileLock.hpp>
+
+#include <errno.h>
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <fcntl.h>
+
+#include <set>
+#include <vector>
+
+#include <core/Algorithm.hpp>
+#include <core/Thread.hpp>
+#include <core/Error.hpp>
+#include <core/Log.hpp>
+#include <core/FilePath.hpp>
+#include <core/FileSerializer.hpp>
+
+#include <boost/foreach.hpp>
+
+namespace rstudio {
+namespace core {
+
+namespace {
+
+std::string makePidString()
+{
```

Comment:
We can use safe_convert::numberToString here. Alternatively `boost::format` / `boost::str` also gives a nice compact notation.